src/test/ui/lint/unused-braces-while-let-with-mutable-value.rs

```diff
@@ -0,0 +1,12 @@
+// check-pass
+
+#![deny(unused_braces)]
+
+fn main() {
+    let mut a = Some(3);
+    // Shouldn't warn below `a`.
+    while let Some(ref mut v) = {a} {
```

Comment:
So it is not linted for unnamed struct literal in the future?

Reply:
not quite sure what you mean here, we can't warn for `{ a }` if it's used in patterns because it can
actually change the behavior there. From the docs
```rust
    /// Due to `ref` pattern, there can be a difference between using
    /// `{ expr }` and `expr` in pattern-matching contexts. This means
    /// that we should only lint `unused_parens` and not `unused_braces`
    /// in this case.
    ///
    /// ```rust
    /// let mut a = 7;
    /// let ref b = { a }; // We actually borrow a copy of `a` here.
    /// a += 1; // By mutating `a` we invalidate any borrows of `a`.
    /// assert_eq!(b + 1, a); // `b` does not borrow `a`, so we can still use it here.
    /// ```
```

Reply:
Playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d28588845e2466227a28572bb6929b30

What I trying to say is that currently we don't allow struct literals in if, while-let. 
```rust
struct Foo {
     a: u32
}

fn foo() {
     while let Foo { a } = Foo { a: 42} {
          ()
     }
}
```

By enabling around `{ a }`, unnamed structs in the future might be hard to to used/migrate.
I am not against this changes, also am not authorative. 
What I want is improving the precision of the lint around `{( yield )}`.

Reply:
So you want forbid `while let pat = { a }`?

I still can't completely follow you here.

Reply:
Yeah, I think it kind of ambiguous in parsing. But I'm not authoritative here.
so let's cc parser devs like @petrochenkov @estebank 

I'm sorry for causing this noise.

Reply:
I would be very uneasy at allowing `while let pat = { a } { /**/ }` precisely due to @lzutao's point. 🤔

Reply:
We already allow `while let pat = { a } { /**/ }` so I don't really think we can change that now

Reply:
But with the unused braces warning people will feel discouraged to use it.

Reply:
`while let pat = { a }` also doesn't cause ambiguity here as `while let _ =` expects `expr {` so the `{` of `{ a }` must be part of `expr` here

Reply:
I'm kinda get it your point. Does it mean parsing unnamed structs at expr is unambiguous and allowed? 
Even though struct literals are forbidden.

Reply:
> But with the unused braces warning people will feel discouraged to use it.

The unused braces warning can break code here and imo it's quite bad for rustc to emit incorrect errors